chore(ecs): describe job and generate run task request from job workload

Lou1415926 opened the pull request:
This PR implement the function to generate a run task request from a deployed job, and the functions to support it.

This PR is rebased from #2201 (which is not merged yet). The real changes to this PR are at 
- `ecs/run_task_request.go` (where the information needed to generated a `task run` command is grabbed from a Copilot job)
- `ecs/run_task_request_test.go`
- `aws/stepfunctions/*` (where the client to make API calls to stepfunctions is implemented)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Lou1415926 committed:
add stepmachine client

Lou1415926 committed:
add stepmachine client

Lou1415926 committed:
describe job neetwork config

Lou1415926 committed:
describe job neetwork config

Lou1415926 committed:
run task request from job

Lou1415926 committed:
run task request from job

Lou1415926 committed:
implement custom unmarshaling to avoid deeply nested struct

Lou1415926 committed:
implement custom unmarshaling to avoid deeply nested struct

bvtujo commented on internal/pkg/cli/flag.go:
```diff
@@ -224,7 +225,11 @@ Must be either "MySQL" or "PostgreSQL".`
+	taskImageTagFlagDescription    = `Optional. The container image tag in addition to "latest".`
+	generateCommandFlagDescription = `Optional. Generate a command with a pre-filled value for each flag.
+To use a non-Copilot ECS service, specify --generate-cmd <cluster name>/<service name>.
+To use a Copilot service, specify --generate-cmd <application>/<environment>/<service>.
```
nit: can we update this language to include jobs?

Lou1415926 replied:
Will address in next PR!

Lou1415926 committed:
fix lint

Lou1415926 committed:
fix lint

bvtujo commented on internal/pkg/cli/task_run.go:
```diff
@@ -481,6 +498,64 @@ func (o *runTaskOpts) Execute() error {
+		}
+	case 3:
+		appName, envName, serviceName := parts[0], parts[1], parts[2]
+		cmd, err = ecs.RunTaskRequestFromService(ecs.New(sess), appName, envName, serviceName)
```
```suggestion
		cmd, err = ecs.RunTaskRequestFromWorkload(ecs.New(sess), appName, envName, workloadName)
```

Lou1415926 replied:
Will address in next PR!

bvtujo commented on internal/pkg/ecs/ecs.go:
```diff
@@ -220,6 +230,55 @@ func (c Client) NetworkConfiguration(app, env, svc string) (*ecs.NetworkConfigur
+
+type NetworkConfiguration ecs.NetworkConfiguration
+
+func (n *NetworkConfiguration) UnmarshalJSON(b []byte) error {
```
Can we edit this comment to make sure that it clarifies what kind of byte stream we expect? Because this unmarshaling logic is quite deeply nested we should be careful to clarify that this unmarshals *just* the network configuration from inside the DescribeStateMachineDefinition response.

Lou1415926 replied:
done!

huanjani commented on internal/pkg/cli/flag.go:
```diff
@@ -224,7 +225,11 @@ Must be either "MySQL" or "PostgreSQL".`
-	taskImageTagFlagDescription = `Optional. The container image tag in addition to "latest".`
+	taskImageTagFlagDescription    = `Optional. The container image tag in addition to "latest".`
+	generateCommandFlagDescription = `Optional. Generate a command with a pre-filled value for each flag.
+To use a non-Copilot ECS service, specify --generate-cmd <cluster name>/<service name>.
```
Q: we have fully surfaced the term "workload" to users at this point, right? If so, use "workload" here and in the next line instead of "service"?

Lou1415926 replied:
Good catch. I should update the flag description in this PR. I'd refrain from using the keyword `workload` because it doesn't appear anywhere formally in our doc site. I will just say "service or job".

The changes to `flag.go` in this PR is a result of #2201. I will rebase this branch to mainline once that one is merged. 
The real changes in this PR are located at `ecs/run_task_request.go` and `aws/stepfunctions`.

huanjani commented on internal/pkg/cli/task_run.go:
```diff
@@ -481,6 +498,64 @@ func (o *runTaskOpts) Execute() error {
+			return nil, fmt.Errorf("generate task run command from service: %w", err)
+		}
+	default:
+		return nil, errors.New("invalid input to --generate-cmd: must be of the form <cluster>/<service> or <app>/<env>/<workload>")
```
same as above: "workload" instead of "service" for the non-Copilot part of the error message?

Lou1415926 committed:
add test

Lou1415926 committed:
add test

Lou1415926 committed:
add public comment

Lou1415926 committed:
add public comment

Lou1415926 committed:
chore: implement ask for secret init (#2271)

Proceeding PR: #2266 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

toricls committed:
docs(ja): add localized docs for Copilot v1.6.0 (#2286)

<!-- Provide summary of changes -->

This PR is a continued work from #2225 by @efekarakus, and adds localized docs for Japanese language in our docs 🚀

Note that the localized docs are translated from the v1.6 Copilot docs - https://github.com/aws/copilot-cli/tree/v1.6.0/site, and are not from the mainline branch.

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

dependabot[bot] committed:
chore: Bump github.com/onsi/gomega from 1.11.0 to 1.12.0 (#2292)

Bumps [github.com/onsi/gomega](https://github.com/onsi/gomega) from 1.11.0 to 1.12.0.
- [Release notes](https://github.com/onsi/gomega/releases)
- [Changelog](https://github.com/onsi/gomega/blob/master/CHANGELOG.md)
- [Commits](https://github.com/onsi/gomega/compare/v1.11.0...v1.12.0)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
chore: Bump github.com/aws/aws-sdk-go from 1.38.30 to 1.38.36 (#2290)

Bumps [github.com/aws/aws-sdk-go](https://github.com/aws/aws-sdk-go) from 1.38.30 to 1.38.36.
- [Release notes](https://github.com/aws/aws-sdk-go/releases)
- [Changelog](https://github.com/aws/aws-sdk-go/blob/main/CHANGELOG.md)
- [Commits](https://github.com/aws/aws-sdk-go/compare/v1.38.30...v1.38.36)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

bvtujo commented on internal/pkg/ecs/ecs.go:
```diff
@@ -47,15 +55,17 @@ type ServiceDesc struct {
-	ecsClient ecsClient
+	rgGetter       resourceGetter
+	ecsClient      ecsClient
+	StepFuncClient stepFunctionsClient
```
Does this client need to be exported?

Lou1415926 replied:
They don't since they will be initialized by `New(sess)` so that the caller do not need to initialize these fields individually across package.

Although I think we could probably refactor this such that the clients are initialized from the caller (seems like a better approach to me) - in which case they will need to be exported.

bvtujo reviewed (commented):
This is looking really good! Some nits about the custom unmarshaler for that deeply nested struct, however!

huanjani reviewed (approved):
Looks good!

Lou1415926 committed:
update custom UnmarshalJSON comment

Lou1415926 committed:
fix conflict
